### PR TITLE
Add ability to configure alembic_version table in DialectImpl

### PR DESCRIPTION
**Pull request by maver1ck**
### Description
This is followup PR to discusion about custom alembic_version table definition
Reference: https://github.com/sqlalchemy/alembic/discussions/1562
Fixes: https://github.com/sqlalchemy/alembic/issues/1560

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by zzzeek** [comment]
ok, nice job with the test.  I can clean that up to make it use a unique name

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 0dc0bda1e8c5182d2d70360e779564a3c44d9097 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 0dc0bda1e8c5182d2d70360e779564a3c44d9097: https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Comment by maver1ck** [comment]
@zzzeek 
I fixed linting issues and hopefully problem with sqla 1.3.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e98a15658ded31eebc75b23be35b3da5d4c631e7 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset e98a15658ded31eebc75b23be35b3da5d4c631e7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 3ff240fc43e2c90cd5d2bb35783aa9d0aa602fd7 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 3ff240fc43e2c90cd5d2bb35783aa9d0aa602fd7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Comment by maver1ck** [comment]
@zzzeek 
I can't find a way of selecting one column that is both compatible with SQLAlchemy 1.3 and (1.4, 2.0).
Any ideas ?

**Comment by CaselIT** [comment]
There is a select you can use in the util sqla_compat module

**Comment by maver1ck** [comment]
@CaselIT 
Fixed.

**Review by sqla-tester — COMMENTED** [review]
Hi, this is **sqla-tester** and I see you've pinged me for review. However, user **maver1ck** is not authorized to initiate CI jobs.  Please wait for a project member to do this!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 14539e42a63031b2459f0ffd8d969b9768d31f96 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 14539e42a63031b2459f0ffd8d969b9768d31f96 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Comment by CaselIT** [comment]
> @CaselIT Fixed. May I ask for another test run ?

sure! thanks for the effort!

**Comment by CaselIT** [comment]
oh I guess since the last CI mypy was updated and you had to fix also mypy to make the ci pass.
Sorry about that.

Thanks and feel free to just tell us to complete this. making the CI happy can be tedious at times

**Comment by maver1ck** [comment]
@CaselIT 
No problem at all. I'm really missing bash script that will run all checks and tests :)

One more test please.

**Comment by CaselIT** [comment]
> @CaselIT No problem at all. I'm really missing bash script that will run all checks and tests :)
> 
> One more test please.

there should be pre-commit configured that can run all tests

File: alembic/ddl/base.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ def __init__(
         self.comment = comment
 
 
-@compiles(RenameTable)  # type: ignore[misc]
```
Comment (zzzeek) [inline alembic/ddl/base.py:178]: oh....OK dont do unrelated typing changes here, if CI is not passing due to new mypy versions we would want to commit a separate fix for just typing issues.   we will do this in gerrit.  dont worry about jenkins not fully passing right now

File: alembic/ddl/impl.py [hunk 2]
```diff
@@ -136,6 +139,32 @@ def static_output(self, text: str) -> None:
         self.output_buffer.write(text + "\n\n")
         self.output_buffer.flush()
 
+    def version_table_impl(
```
Comment (zzzeek) [inline alembic/ddl/impl.py:142]: I think we should make this more future proof by doing all keyword args with room for expansion.  these are keyword only arguments:

```
def version_table_impl(
   self, *, version_table, version_table_schema, version_table_pk, **kw
)
```

File: alembic/runtime/migration.py [hunk 3]
```diff
@@ -212,6 +197,13 @@ def __init__(
             self.output_buffer,
             opts,
         )
+
```
Comment (zzzeek) [inline alembic/runtime/migration.py:200]: so then here:

```
 self._version = self.impl.version_table_impl(
            version_table=version_table,
            version_table_schema=version_table_schema,
            version_table_pk=opts.get("version_table_pk", True),
        )
```

**Review by zzzeek — CHANGES_REQUESTED** [review]

File: alembic/ddl/impl.py (hunk 2 above)
Comment (maver1ck) [reply alembic/ddl/impl.py:142]: Fixed

**Comment by maver1ck** [comment]
> > @CaselIT No problem at all. I'm really missing bash script that will run all checks and tests :)
> > One more test please.
> 
> there should be pre-commit configured that can run all tests

pre-commit is not triggering tests and mypy :(

**Comment by maver1ck** [comment]
@zzzeek 
Please also find PR fixing mypy issues. https://github.com/sqlalchemy/alembic/pull/1564

**Comment by CaselIT** [comment]
@maver1ck can you rebase / merge main?

**Comment by maver1ck** [comment]
> @maver1ck can you rebase / merge main?

Rebased.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e70fdc8f4e405cabf5099c2100763d7b24da3be8 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset e70fdc8f4e405cabf5099c2100763d7b24da3be8 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Review by zzzeek — APPROVED** [review]
approved, we will add some final elements in the gerrit side thanks!

File: alembic/ddl/impl.py [hunk 4]
```diff
@@ -136,6 +139,32 @@ def static_output(self, text: str) -> None:
...
+        self, *, version_table, version_table_schema, version_table_pk, **kw
```
Comment (sqla-tester) [inline alembic/ddl/impl.py:143]: **Federico Caselli** (CaselIT) wrote:

missing typing. Should likely be:
```
 version_table: str, version_table_schema: Optional[str], version_table_pk: bool, **kw: Any
```

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

File: alembic/ddl/impl.py [hunk 5]
```diff
@@ -136,6 +139,32 @@
...
+    ) -> Table:
+        """create the Table object for the version_table.
+
+        Provided as part of impl so that third party dialects can override
+        this.
+
+        .. versionadded:: 1.13.4
+
+        """
+        vt = Table(
```
Comment (sqla-tester) [inline alembic/ddl/impl.py:153]: **Federico Caselli** (CaselIT) wrote:

it should 1.14

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

I've left a few comment, nothing major though.

Great work!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

* /COMMIT_MSG (line 18): we should add a line that says
```
Fixes: #1560
```
* docs/build/unreleased/vers.rst (line 2): missing ticket here. also the file name could be named as the ticked.

Issue is 1560

File: alembic/ddl/impl.py (hunk 4 above)
Comment (maver1ck) [reply alembic/ddl/impl.py:143]: Fixed.

File: alembic/ddl/impl.py (hunk 5 above)
Comment (maver1ck) [reply alembic/ddl/impl.py:153]: Done

**Comment by maver1ck** [comment]
@CaselIT 
I fixed typing issues in the code however I don't know how to handle commit message and vers.rst file

File: alembic/ddl/impl.py [hunk 6]
```diff
@@ -136,6 +139,37 @@ def static_output(self, text: str) -> None:
...
+        *,
+        version_table: str,
+        version_table_schema: Optional[str],
+        version_table_pk: bool,
+        **kw,
+    ) -> Table:
+        """create the Table object for the version_table.
+
+        Provided as part of impl so that third party dialects can override
+        this.
```
Comment (sqla-tester) [inline alembic/ddl/impl.py:153]: **Michael Bayer** (zzzeek) wrote:

im not convinced we need to bump here, but I guess numbers are free

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

* docs/build/unreleased/vers.rst (line 2): I repurposed the one opened for the initial PR. Seemed like it fit

File: alembic/ddl/impl.py [hunk 7]
```diff
@@ -136,6 +139,37 @@ def static_output(self, text: str) -> None:
         self.output_buffer.write(text + "\n\n")
         self.output_buffer.flush()
 
+    def version_table_impl(
+        self,
```
Comment (sqla-tester) [inline alembic/ddl/impl.py:143]: **Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

* /COMMIT_MSG (line 18): Done
* docs/build/unreleased/vers.rst (line 2): Done

File: alembic/ddl/impl.py (hunk 6 above)
Comment (sqla-tester) [reply alembic/ddl/impl.py:153]: **Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/5556 has been **merged**. Congratulations! :)

**Pull request closed without merging**